### PR TITLE
Report mAP with cuhk_anet17_pred

### PR DESCRIPTION
**Pull request by kennymckormick**

### CODE REVIEW

File: configs/localization/bmn/README.md [hunk 1]
```diff
@@ -9,23 +9,36 @@
...
 
-Notes:
 1. The **gpus** indicates the number of gpu we used to get the checkpoint.
 According to the [Linear Scaling Rule](https://arxiv.org/abs/1706.02677), you may set the learning rate proportional to the batch size if you use different GPUs or videos per GPU,
 e.g., lr=0.01 for 4 GPUs * 2 video/gpu and lr=0.08 for 16 GPUs * 4 video/gpu.
 2. For feature column, cuhk_mean_100 denotes the widely used cuhk activitynet feature extracted by [anet2016-cuhk](https://github.com/yjxiong/anet2016-cuhk), mmaction_video and mmaction_clip denote feature extracted by mmaction, with video-level activitynet finetuned model or clip-level activitynet finetuned model respectively.
+3. We evaluate the action detection performance of BMN, using  [anet_cuhk_2017](https://openmmlab.oss-accelerate.aliyuncs.com/mmaction/localization/cuhk_anet17_pred.json) submission for ActivityNet2017 Untrimmed Video Classification Track to assign label for each action proposal.
+
+*We train BMN with the official repo, evaluate its proposal generation and action detection performance with [anet_cuhk_2017](https://openmmlab.oss-accelerate.aliyuncs.com/mmaction/localization/cuhk_anet17_pred.json) for label assigning.
```
Comment (innerlee) [inline configs/localization/bmn/README.md:41]: `official repo` refers to ?

File: mmaction/core/evaluation/eval_detection.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
+import json
+
+import numpy as np
+import pandas as pd
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:4]: `pandas` is a heavy package, we may need to implement it by ourselves.

**Comment by dreamerlin** [comment]
It seems the docstring is not consistent with the format.

File: mmaction/core/evaluation/accuracy.py [hunk 3]
```diff
@@ -356,12 +356,6 @@ def get_weighted_score(score_list, coeff_list):
     return weighted_scores
 
 
-def softmax(x, dim=1):
-    """Compute softmax values for each sets of scores in x."""
-    e_x = np.exp(x - np.max(x, axis=dim, keepdims=True))
-    return e_x / e_x.sum(axis=dim, keepdims=True)
-
-
```
Comment (dreamerlin) [inline mmaction/core/evaluation/accuracy.py:364]: Why delete it?

File: mmaction/core/evaluation/eval_detection.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+class ANETdetection(object):
+
+    def __init__(self,
+                 ground_truth_filename=None,
+                 prediction_filename=None,
+                 tiou_thresholds=np.linspace(0.5, 0.95, 10),
+                 verbose=False):
+        if not ground_truth_filename:
+            raise IOError('Please input a valid ground truth file.')
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:18]: FileNotFoundError

File: mmaction/core/evaluation/eval_detection.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        self.ground_truth, self.activity_index = self._import_ground_truth(
+            ground_truth_filename)
+        self.prediction = self._import_prediction(prediction_filename)
+
+        if self.verbose:
+            print('[INIT] Loaded ground_truth from '
+                  f'{self.ground_truth_filename}, prediction from '
+                  f'{self.prediction_filename}.')
+            nr_gt = len(self.ground_truth)
+            print('\tNumber of ground truth instances: {}'.format(nr_gt))
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:36]: f-string

File: mmaction/core/evaluation/eval_detection.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        # Import ground truth and predictions.
+        self.ground_truth, self.activity_index = self._import_ground_truth(
+            ground_truth_filename)
+        self.prediction = self._import_prediction(prediction_filename)
+
+        if self.verbose:
+            print('[INIT] Loaded ground_truth from '
+                  f'{self.ground_truth_filename}, prediction from '
+                  f'{self.prediction_filename}.')
+            nr_gt = len(self.ground_truth)
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:35]: `num_gts`

File: mmaction/core/evaluation/eval_detection.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            print('\tNumber of ground truth instances: {}'.format(nr_gt))
+            nr_pred = len(self.prediction)
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:37]: num_preds

File: mmaction/core/evaluation/eval_detection.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+            Full path to the ground truth json file.
+
+        Outputs
+        -------
+        ground_truth : df
+            Data frame containing the ground truth instances.
+        activity_index : dict
+            Dictionary containing class index.
+        """
+        with open(ground_truth_filename, 'r') as fobj:
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:58]: fobj -> f

File: mmaction/core/evaluation/accuracy.py [hunk 9]
```diff
@@ -465,3 +459,42 @@ def average_precision_at_temporal_iou(ground_truth,
...
+    tt1 = np.maximum(target_segment[0], candidate_segments[:, 0])
+    tt2 = np.minimum(target_segment[1], candidate_segments[:, 1])
+    # Intersection including Non-negative overlap score.
+    segments_intersection = (tt2 - tt1).clip(0)
+    # Segment union.
+    segments_union = (candidate_segments[:, 1] - candidate_segments[:, 0]) \
+        + (target_segment[1] - target_segment[0]) - segments_intersection
+    # Compute overlap as the ratio of the intersection
+    # over union of two segments.
+    tIoU = segments_intersection.astype(float) / segments_union
```
Comment (dreamerlin) [inline mmaction/core/evaluation/accuracy.py:487]: tiou

Comment (dreamerlin) [reply mmaction/core/evaluation/accuracy.py:487]: @tony2016uestc Is this function redundant ?

File: mmaction/core/evaluation/eval_detection.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+            data = json.load(fobj)
+        # Checking format
+        activity_index, cidx = {}, 0
+        video_lst, t_start_lst, t_end_lst, label_lst = [], [], [], []
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:62]: lst -> list

File: mmaction/core/evaluation/eval_detection.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+        for videoid, v in data.items():
+            for ann in v['annotations']:
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:64]: ann -> anno

File: mmaction/core/evaluation/eval_detection.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        -------
+        ground_truth : df
+            Data frame containing the ground truth instances.
+        activity_index : dict
+            Dictionary containing class index.
+        """
+        with open(ground_truth_filename, 'r') as fobj:
+            data = json.load(fobj)
+        # Checking format
+        activity_index, cidx = {}, 0
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:61]: cidx -> class_idx

File: mmaction/core/evaluation/eval_detection.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+        video_lst, t_start_lst, t_end_lst, label_lst = [], [], [], []
+        for videoid, v in data.items():
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:63]: video_id, video_info

File: mmaction/core/evaluation/eval_detection.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+        ----------
+        prediction_filename : str
+            Full path to the prediction json file.
+
+        Outputs
+        -------
+        prediction : df
+            Data frame containing the prediction instances.
+        """
+        with open(prediction_filename, 'r') as fobj:
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:95]: fobj -> f

File: mmaction/core/evaluation/eval_detection.py [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+    def _get_predictions_with_label(self, prediction_by_label, label_name,
+                                    cidx):
+        """Get all predicitons of the given label.
+
+        Return empty DataFrame if there is no predcitions with the given label.
+        """
+        if cidx in prediction_by_label.groups.keys():
+            return prediction_by_label.get_group(cidx).reset_index(drop=True)
+        else:
+            print('Warning: No predictions of label \'%s\' were provdied.' %
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:126]: double-quota for the outer

File: mmaction/core/evaluation/eval_detection.py [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+        })
+        return prediction
+
+    def _get_predictions_with_label(self, prediction_by_label, label_name,
+                                    cidx):
+        """Get all predicitons of the given label.
+
+        Return empty DataFrame if there is no predcitions with the given label.
+        """
+        if cidx in prediction_by_label.groups.keys():
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:123]: for class_idx in prediction_by_label

File: mmaction/core/evaluation/eval_detection.py [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+        prediction_by_label = self.prediction.groupby('label')
+
+        results = Parallel(n_jobs=len(self.activity_index))(
+            delayed(compute_average_precision_detection)(
+                ground_truth=ground_truth_by_label.get_group(cidx).reset_index(
+                    drop=True),
+                prediction=self._get_predictions_with_label(
+                    prediction_by_label, label_name, cidx),
+                tiou_thresholds=self.tiou_thresholds,
+            ) for label_name, cidx in list(self.activity_index.items()))
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:145]: for label_name, class_idx in self.activity_index.items()

File: mmaction/core/evaluation/eval_detection.py [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        Required fields: ['video-id, 't-start', 't-end', 'score']
+    tiou_thresholds : 1darray, optional
+        Temporal intersection over union threshold.
+
+    Outputs
+    -------
+    ap : float
+        Average precision score.
+    """
+    ap = np.zeros(len(tiou_thresholds))
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:191]: zeros_like?

File: mmaction/core/evaluation/eval_detection.py [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+    if prediction.empty:
+        return ap
+
+    npos = float(len(ground_truth))
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:195]: num_positive

File: mmaction/core/evaluation/eval_detection.py (hunk 18 above)
Comment (dreamerlin) [reply mmaction/core/evaluation/eval_detection.py:191]:
```
num_thresholds = len(tiou_thresholds)
num_positives = len(ground_truth)
num_predictions = len(prediction)
```

File: mmaction/core/evaluation/accuracy.py (hunk 3 above)
Comment (dreamerlin) [reply mmaction/core/evaluation/accuracy.py:364]: Ok, it seem it is the same with `from scipy.special import softmax`

File: tools/analysis/report_map.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+def cuhk17_top1():
+    global args
+    if not osp.exists('cuhk_anet17_pred.json'):
+        os.system('wget https://openmmlab.oss-accelerate.aliyuncs.com/'
+                  'mmaction/localization/cuhk_anet17_pred.json')
+    proposal = mmcv.load(args.proposal)
+    results = proposal['results']
+    cuhk = mmcv.load('cuhk_anet17_pred.json')
+    cuhk = cuhk['results']
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:21]: `cuhk_pred = mmcv.load('cuhk_anet17_pred.json')['results']`

File: tools/analysis/report_map.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+    func()
+    anet_detection = ANETdetection(
+        args.gt,
+        args.det_output,
+        tiou_thresholds=np.linspace(0.5, 0.95, 10),
+        verbose=True)
+    mAP, average_mAP = anet_detection.evaluate()
+    print('[RESULTS] Performance on ActivityNet detection task.')
+    print(f'mAP: {mAP}')
+    print(f'Average-mAP: {average_mAP}')
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:83]:
```
print('[RESULTS] Performance on ActivityNet detection task.\n'
        f'mAP: {mAP}\nAverage-mAP: {average_mAP}')
```

File: tools/analysis/report_map.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+
+    def get_topk(preds, k):
+        preds.sort(key=lambda x: x['score'])
+        return preds[-k:]
+
+    for k, v in results.items():
+        action_pred = cuhk[k]
+        top1 = get_topk(action_pred, 1)
+        lb = top1[0]['label']
+        newv = []
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:31]: `newv` -> `new_value`

File: tools/analysis/report_map.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+    cuhk = cuhk['results']
+
+    def get_topk(preds, k):
+        preds.sort(key=lambda x: x['score'])
+        return preds[-k:]
+
+    for k, v in results.items():
+        action_pred = cuhk[k]
+        top1 = get_topk(action_pred, 1)
+        lb = top1[0]['label']
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:30]: `lb` -> top1_label

File: tools/analysis/report_map.py [hunk 24]
```diff
@@ -0,0 +1,87 @@
+import argparse
+import os
+import os.path as osp
+
+import mmcv
+import numpy as np
+
+from mmaction.core import ANETdetection
+
+args = None
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:10]: It seems this definition is redundant

File: mmaction/core/evaluation/eval_detection.py [hunk 25]
```diff
@@ -0,0 +1,249 @@
+import json
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+
+from .accuracy import interpolated_prec_rec, segment_iou
+
+
+class ANETdetection(object):
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:8]: `ANetDection`

File: mmaction/core/evaluation/eval_detection.py [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+        ground_truth_by_label = self.ground_truth.groupby('label')
+        prediction_by_label = self.prediction.groupby('label')
+
+        results = Parallel(n_jobs=len(self.activity_index))(
+            delayed(compute_average_precision_detection)(
+                ground_truth=ground_truth_by_label.get_group(cidx).reset_index(
+                    drop=True),
+                prediction=self._get_predictions_with_label(
+                    prediction_by_label, label_name, cidx),
+                tiou_thresholds=self.tiou_thresholds,
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py]: the last comma should be removed.

File: mmaction/core/evaluation/eval_detection.py [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+    # Sort predictions by decreasing score order.
+    sort_idx = prediction['score'].values.argsort()[::-1]
+    prediction = prediction.loc[sort_idx].reset_index(drop=True)
+
+    # Initialize true positive and false positive vectors.
+    tp = np.zeros((len(tiou_thresholds), len(prediction)))
+    fp = np.zeros((len(tiou_thresholds), len(prediction)))
+
+    # Adaptation to query faster
+    ground_truth_gbvn = ground_truth.groupby('video-id')
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py]: What does `gbvn` mean ?

File: tools/analysis/report_map.py [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def cuhk17_top1():
+    global args
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:14]: for a global variable, `global args` can only be written in `def main():` once.

File: configs/localization/bmn/README.md (hunk 1 above)
Comment (kennymckormick) [reply configs/localization/bmn/README.md:41]: BMN official repo, link added

File: mmaction/core/evaluation/accuracy.py (hunk 3 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/accuracy.py]: ... seems scipy is not an allowed package, roll back this commit

File: mmaction/core/evaluation/eval_detection.py (hunk 4 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/eval_detection.py:16]: Guess FileNotFoundError should be used when `not os.path.exists(filename)`. Here we just check if the user provides a value as the argument

File: mmaction/core/evaluation/eval_detection.py (hunk 27 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/eval_detection.py]: group by video names, now the code is updated

File: mmaction/core/evaluation/eval_detection.py (hunk 15 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/eval_detection.py]: code removed

File: mmaction/core/evaluation/eval_detection.py (hunk 16 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/eval_detection.py]: code removed

File: mmaction/core/evaluation/eval_detection.py (hunk 17 above)
Comment (kennymckormick) [reply mmaction/core/evaluation/eval_detection.py]: code updated

File: mmaction/core/evaluation/accuracy.py [hunk 29]
```diff
@@ -465,3 +465,42 @@ def average_precision_at_temporal_iou(ground_truth,
                                                   recall_cumsum[t_idx, :])
 
     return ap
+
+
+def segment_iou(target_segment, candidate_segments):
```
Comment (dreamerlin) [inline mmaction/core/evaluation/accuracy.py:470]: Duplicated with the existing function `pairwise_temporal_iou`.

File: mmaction/core/evaluation/eval_detection.py [hunk 30]
```diff
@@ -0,0 +1,246 @@
+import json
+
+import numpy as np
+
+from .accuracy import interpolated_precision_recall, pairwise_temporal_iou
+
+
+class ANetDetection(object):
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:8]: Docstring is needed here.

File: mmaction/core/evaluation/eval_detection.py [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+                label_list.append(label)
+                score_list.append(result['score'])
+        prediction = [{
+            'video-id': vid,
+            't-start': tst,
+            't-end': ted,
+            'label': lb,
+            'score': sc
+        } for vid, tst, ted, lb, sc in zip(video_list, t_start_list,
+                                           t_end_list, label_list, score_list)]
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:114]:
```
prediction = []
for video_id, video_info in data['results'].items():
    for result in video_info:
        label = self.activity_index[result['label']]
        prediction_item = dict(
            video-id=video_id,
            t-start=float(result['segment'][0]),
            t-end=float(result['segment'][1]),
            label=label,
            score=result['score'])
        prediction.append(prediction_item)
```

File: mmaction/core/evaluation/eval_detection.py [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+                t_end_list.append(float(anno['segment'][1]))
+                label_list.append(activity_index[anno['label']])
+
+        ground_truth = [{
+            'video-id': vid,
+            't-start': tst,
+            't-end': ted,
+            'label': lb
+        } for vid, tst, ted, lb in zip(video_list, t_start_list, t_end_list,
+                                       label_list)]
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:77]:
```
ground_truth = []
activity_indexes = {}
for video_id, video_info in data.items():
    for anno in video_info['annotations']:
        label = anno['label']
        if label not in activity_index:
            activity_index[label] = len(activity_index)
        gt_item = dict(
            video-id=video_id[2:],
            t-start=float(anno['segment'][0]),
            t-end=float(anno['segment'][1]),
            label=activity_index[label])
        ground_truth.append(gt_item)
```

File: mmaction/core/evaluation/eval_detection.py [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+        ground_truth_filename : str
+            Full path to the ground truth json file.
+
+        Outputs
+        -------
+        ground_truth : list
+            List containing the ground truth instances (dictionaries).
+        activity_index : dict
+            Dictionary containing class index.
+        """
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:54]:
```
"""This function reads ground truth file, returns the ground truth instances and the activity classes.

Args:
    ground_truth_filename (str): Full path to the ground truth json file.

Returns:
    tuple[list, dict]: (ground_truth, activity_index).
        ground_truth contains the ground truth instances, which is in a dict format.
        activity_index contains classes index.
"""
```

File: mmaction/core/evaluation/eval_detection.py [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+        results = [
+            compute_average_precision_detection(
+                ground_truth=ground_truth_by_label[i],
+                prediction=prediction_by_label[i],
+                tiou_thresholds=self.tiou_thresholds)
+            for i in range(len(self.activity_index))
+        ]
+
+        for i in range(len(self.activity_index)):
+            ap[:, i] = results[i]
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:141]:
```
for i in range(len(self.activity_index)):
    ap_result = compute_average_precision_detection(ground_truth_by_label[i], 
                                                    prediction_by_label[i], 
                                                    self.tiou_thresholds)
    ap[:, i] = ap_result 
```

File: mmaction/core/evaluation/eval_detection.py [hunk 35]
```diff
@@ -0,0 +1,246 @@
...
+    num_positive = float(num_gts)
+    lock_gt = np.ones((num_thresholds, num_gts)) * -1
+    # Sort predictions by decreasing score order.
+    prediction.sort(key=lambda x: -x['score'])
+    # Initialize true positive and false positive vectors.
+    tp = np.zeros((num_thresholds, num_preds))
+    fp = np.zeros((num_thresholds, num_preds))
+
+    # Adaptation to query faster
+    ground_truth_gbvn = {}
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:200]: `gbvn` is ambiguous

File: mmaction/core/evaluation/eval_detection.py [hunk 36]
```diff
@@ -0,0 +1,246 @@
...
+
+        if self.verbose:
+            print('[INIT] Loaded ground_truth from '
+                  f'{self.ground_truth_filename}, prediction from '
+                  f'{self.prediction_filename}.')
+            num_gts = len(self.ground_truth)
+            print(f'Number of ground truth instances: {num_gts}')
+            num_preds = len(self.prediction)
+            print(f'Number of predictions: {num_preds}')
+            print(f'Fixed threshold for tiou score: {self.tiou_thresholds}')
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:37]:
```
from mmcv.utils import get_root_logger, print_log
...
self.logger = get_root_logger()
log_msg = '[INIT] Loaded ground_truth from '
    f'{self.ground_truth_filename}, prediction from '
    f'{self.prediction_filename}.\n'
    f'Number of ground truth instances: {len(self.ground_truth)}\n'
    f'Number of predictions: {len(self.prediction)}\n'
    f'Fixed threshold for tiou score: {self.tiou_thresholds}'
print_log(msg, logger=logger)
```

File: mmaction/core/evaluation/eval_detection.py [hunk 37]
```diff
@@ -0,0 +1,246 @@
...
+    tp = np.zeros((num_thresholds, num_preds))
+    fp = np.zeros((num_thresholds, num_preds))
+
+    # Adaptation to query faster
+    ground_truth_gbvn = {}
+    for i, item in enumerate(ground_truth):
+        item['index'] = i
+        if item['video-id'] not in ground_truth_gbvn:
+            ground_truth_gbvn[item['video-id']] = []
+        ground_truth_gbvn[item['video-id']].append(item)
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:205]:
```
ground_truth_gbvn.setdefault(item['video-id'], []).append(item)
```

File: mmaction/core/evaluation/eval_detection.py [hunk 38]
```diff
@@ -0,0 +1,246 @@
...
+            fp[:, idx] = 1
+            continue
+
+        tiou_arr = pairwise_temporal_iou(
+            np.array([pred['t-start'], pred['t-end']]),
+            np.array([np.array([gt['t-start'], gt['t-end']]) for gt in gts]))
+        tiou_arr = tiou_arr.reshape(-1)
+        # We would like to retrieve the predictions with highest tiou score.
+        tiou_sorted_idx = tiou_arr.argsort()[::-1]
+        for tidx, tiou_thr in enumerate(tiou_thresholds):
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:221]: `tiou_thr` -> `tiou_threshold`

File: mmaction/core/evaluation/eval_detection.py [hunk 39]
```diff
@@ -0,0 +1,246 @@
...
+            for jdx in tiou_sorted_idx:
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:222]: `jdx` -> `j_idx`

File: mmaction/core/evaluation/eval_detection.py [hunk 40]
```diff
@@ -0,0 +1,246 @@
...
+            if fp[tidx, idx] == 0 and tp[tidx, idx] == 0:
+                fp[tidx, idx] = 1
+
+    tp_cumsum = np.cumsum(tp, axis=1).astype(np.float)
+    fp_cumsum = np.cumsum(fp, axis=1).astype(np.float)
+    recall_cumsum = tp_cumsum / num_positive
+
+    precision_cumsum = tp_cumsum / (tp_cumsum + fp_cumsum)
+
+    for tidx in range(len(tiou_thresholds)):
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:242]: `tidx` -> `t_idx`

**Comment by innerlee** [comment]
looks good.
one last comment: is ANet the standard or de facto abbreviation for ActivityNet?

**Comment by kennymckormick** [comment]
> looks good.
> one last comment: is ANet the standard or de facto abbreviation for ActivityNet?
> 
> edit:
> The class name ActivityNetDetection is more clear since there are terms like ONet, PNet etc. in face detection.
> 
> Other usage of anet is okay

In the ActivityNet [official repo](https://github.com/activitynet/ActivityNet) it is abbreviated as ANET.

File: mmaction/core/evaluation/eval_detection.py [hunk 41]
```diff
@@ -0,0 +1,234 @@
+import json
+
+import numpy as np
+from mmcv.utils import get_logger, print_log
+
+from .accuracy import interpolated_precision_recall, pairwise_temporal_iou
+
+
+class ANetDetection(object):
+    """class to Evaluate detection results on ActivityNet.
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:10]: Class to evaluate detection results on ActivityNet.

File: mmaction/core/evaluation/eval_detection.py [hunk 42]
```diff
@@ -0,0 +1,234 @@
...
+
+    Args:
+        ground_truth_filename (str): The filename of groundtruth.
+        prediction_filename (str): The filename of action detection results.
+        tiou_thresholds (np.ndarray): The thresholds of t_iou to evaluate.
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:15]: tiou_thresholds (np.ndarray): The thresholds of temporal iou to evaluate.

File: mmaction/core/evaluation/eval_detection.py [hunk 43]
```diff
@@ -0,0 +1,234 @@
...
+        verbose (bool): Print verbose logs.
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:16]: Whether to print verbose logs.

File: mmaction/core/evaluation/eval_detection.py [hunk 44]
```diff
@@ -0,0 +1,234 @@
...
+        Parameters
+        ----------
+        prediction_filename : str
+            Full path to the prediction json file.
+
+        Outputs
+        -------
+        prediction : list
+            List containing the prediction instances (dictionaries).
+        """
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:95]:
```
"""Read prediction file and return the prediction instances.

Args:
    prediction_filename (str): Full path to the prediction json file.

Returns:
    List: List containing the prediction instances (dictionaries).
"""
```

File: mmaction/core/evaluation/eval_detection.py [hunk 45]
```diff
@@ -0,0 +1,234 @@
...
+        List containing the prediction instances (dictionaries).
+        Required keys: ['video-id, 't-start', 't-end', 'score']
+    tiou_thresholds : 1darray, optional
+        Temporal intersection over union threshold.
+
+    Outputs
+    -------
+    ap : float
+        Average precision score.
+    """
```
Comment (dreamerlin) [inline mmaction/core/evaluation/eval_detection.py:173]:
```
"""Compute average precision (detection task) between ground truth and
predictions data frames. If multiple predictions occurs for the same
predicted segment, only the one with highest score is matches as true
positive. This code is greatly inspired by Pascal VOC devkit.

Args:
    ground_truth (list[dict]): List containing the ground truth instances (dictionaries).
        Required keys are 'video-id', 't-start' and 't-end'.
    prediction (list[dict]): List containing the prediction instances (dictionaries).
        Required keys are: 'video-id', 't-start', 't-end' and 'score'.
    tiou_thresholds (np.ndarray): A 1darray indicates the temporal intersection over union threshold, 
        which is optional. Default: ``np.linspace(0.5, 0.95, 10)``.

Returns:
    Float: ap, Average precision score.
"""
```

File: mmaction/core/evaluation/eval_detection.py (hunk 43 above)
Comment (dreamerlin) [reply mmaction/core/evaluation/eval_detection.py:16]: And remember to add default value for the args in docstring.

File: tools/analysis/report_map.py [hunk 46]
```diff
@@ -0,0 +1,84 @@
...
+
+import mmcv
+import numpy as np
+
+from mmaction.core import ANetDetection
+
+args = None
+
+
+def cuhk17_top1():
```
Comment (dreamerlin) [inline tools/analysis/report_map.py:13]:
```
"Assign label for each proposal with the cuhk17 result"
```

And also add the link for the source for the cuhk17.

**Comment by dreamerlin** [comment]
LGTM, @innerlee

**Comment by dreamerlin** [comment]
Remember to update the `changelog.md`

**Review by innerlee — APPROVED** [review]

**Pull request closed without merging**